src/fundus/scraping/common_crawl/pipeline.py

```diff
@@ -0,0 +1,284 @@
+
+        if start < datetime(2016, 8, 1):
+            raise ValueError("The default, and earliest possible, start date is 2016/08/01.")
+
+        if end > datetime.now():
+            raise ValueError("The specified end date is in the future. We don't want to give spoilers, do we?")
+
+        date_sequence: List[datetime] = list(rrule(MONTHLY, dtstart=start, until=end))
+        urls: List[str] = [
+            f"{self.server_address}crawl-data/CC-NEWS/{date.strftime('%Y/%m')}/warc.paths.gz" for date in date_sequence
+        ]
+
+        def load_paths(url: str) -> List[str]:
+            with requests.Session() as session:
+                return gzip.decompress(session.get(url).content).decode("utf-8").split()
+
+        # running two threads per core
+        max_number_of_threads = 2 * (os.cpu_count() or 1)
+
+        with ThreadPool(processes=min(len(urls), max_number_of_threads)) as pool:
+            warc_paths = more_itertools.flatten(
+                list(
+                    tqdm(pool.imap_unordered(load_paths, urls), total=len(urls), desc="Loading WARC paths", leave=False)
+                )
+            )
+
+        start_strf = start.strftime("%Y%m%d%H%M%S")
+        end_strf = end.strftime("%Y%m%d%H%M%S")
+
+        def filter_warc_path_by_date(path: str) -> bool:
+            match: Optional[re.Match[str]] = date_pattern.search(path)
+            if match is None:
+                raise AssertionError(f"Invalid WARC path {path!r}")
+            return start_strf <= match["date"] <= end_strf
+
+        return sorted(
+            (f"{self.server_address}{warc_path}" for warc_path in filter(filter_warc_path_by_date, warc_paths)),
+            reverse=True,
+        )
+
+    @staticmethod
+    def _fetch_articles(
+        warc_path: str,
+        publishers: Tuple[PublisherEnum],
+        error_handling: Literal["suppress", "catch", "raise"],
+        extraction_filter: Optional[ExtractionFilter] = None,
+        url_filter: Optional[URLFilter] = None,
+    ) -> Iterator[Article]:
+        source = CCNewsSource(*publishers, warc_path=warc_path)
+        scraper = CCNewsScraper(source)
+        yield from scraper.scrape(error_handling, extraction_filter, url_filter)
+
+    @staticmethod
+    def _single_crawl(warc_paths: List[str], article_task: Callable[[str], Iterator[Article]]) -> Iterator[Article]:
+        for warc_path in warc_paths:
+            yield from article_task(warc_path)
+
+    def _parallel_crawl(
+        self, warc_paths: List[str], article_task: Callable[[str], Iterator[Article]]
+    ) -> Iterator[Article]:
+        # As one could think, because we're downloading a bunch of files, this task is IO bound, but it is actually
+        # process bound. The reason for this is that we stream the data and process it on the fly rather than
+        # downloading all files and processing them afterward. Therefore, we utilize multiprocessing here instead
+        # of multithreading.
```

Comment:
Thank you very much for adding the comments (and the following ones). I think this definitely helps. Just some minor spell-checking.

```suggestion
        # As one could think, because we're downloading a bunch of files, this task is IO-bound, but it is actually
        # process-bound. The reason is that we stream the data and process it on the fly rather than downloading all
        # files and processing them afterwards. Therefore, we utilize multiprocessing here instead of multithreading.
```